### PR TITLE
Refer to canonical pacakge in case of duplicates

### LINKED ISSUE #93: Resolution of conflicts between a function and a method that's being reexported
Packages involved: plyr, dplyr, https://github.com/poissonconsulting/mcmcdata

`plyr::rename()` is a function, mcmcdata reexports `dplyr::rename()` .

## Problem

``` r
library(conflicted)
library(plyr)
library(dplyr)
library(mcmcdata)

conflict_prefer("rename", "dplyr", "plyr")
#> [conflicted] Will prefer dplyr::rename over plyr::rename.

rename
#> Error:
#> ! [conflicted] rename found in 2 packages.
#> Either pick the one you want with `::`:
#> • dplyr::rename
#> • mcmcdata::rename
#> Or declare a preference with `conflicts_prefer()`:
#> • `conflicts_prefer(dplyr::rename)`
#> • `conflicts_prefer(mcmcdata::rename)`
#> Backtrace:
#>     ▆
#>  1. └─conflicted (local) `<fn>`()
#>  2.   └─cli::cli_abort(...) at conflicted/R/disambiguate.R:51:4
#>  3.     └─rlang::abort(...)
```

<sup>Created on 2023-09-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

## Workaround

Change the order of attaching packages.

``` r
library(conflicted)
library(plyr)
library(mcmcdata)
library(dplyr)

conflict_prefer("rename", "dplyr", "plyr")
#> [conflicted] Will prefer dplyr::rename over plyr::rename.

rename
#> function (.data, ...) 
#> {
#>     UseMethod("rename")
#> }
#> <bytecode: 0x1122fae60>
#> <environment: namespace:dplyr>
```

<sup>Created on 2023-09-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

## Observation

Other examples:

### dplyr, mcmcdata, plyr

``` r
library(conflicted)
library(dplyr)
library(mcmcdata)
library(plyr)
#> ------------------------------------------------------------------------------
#> You have loaded plyr after dplyr - this is likely to cause problems.
#> If you need functions from both plyr and dplyr, please load plyr first, then dplyr:
#> library(plyr); library(dplyr)
#> ------------------------------------------------------------------------------

rename
#> Error:
#> ! [conflicted] rename found in 2 packages.
#> Either pick the one you want with `::`:
#> • plyr::rename
#> • mcmcdata::rename
#> Or declare a preference with `conflicts_prefer()`:
#> • `conflicts_prefer(plyr::rename)`
#> • `conflicts_prefer(mcmcdata::rename)`
#> Backtrace:
#>     ▆
#>  1. └─conflicted (local) `<fn>`()
#>  2.   └─cli::cli_abort(...) at conflicted/R/disambiguate.R:51:4
#>  3.     └─rlang::abort(...)
```

<sup>Created on 2023-09-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### mcmcdata, dplyr, plyr

``` r
library(conflicted)
library(mcmcdata)
library(dplyr)
library(plyr)
#> ------------------------------------------------------------------------------
#> You have loaded plyr after dplyr - this is likely to cause problems.
#> If you need functions from both plyr and dplyr, please load plyr first, then dplyr:
#> library(plyr); library(dplyr)
#> ------------------------------------------------------------------------------

rename
#> Error:
#> ! [conflicted] rename found in 2 packages.
#> Either pick the one you want with `::`:
#> • plyr::rename
#> • dplyr::rename
#> Or declare a preference with `conflicts_prefer()`:
#> • `conflicts_prefer(plyr::rename)`
#> • `conflicts_prefer(dplyr::rename)`
#> Backtrace:
#>     ▆
#>  1. └─conflicted (local) `<fn>`()
#>  2.   └─cli::cli_abort(...) at conflicted/R/disambiguate.R:51:4
#>  3.     └─rlang::abort(...)
```

<sup>Created on 2023-09-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by krlmlr**
This also deals with the case where two packages import from a third package, and a fourth package overwrites it.

Will add tests.

Closes #93.

## Example

``` r
library(conflicted)
library(plyr)
library(dm)
library(mcmcdata)

mutate
#> Error:
#> ! [conflicted] mutate found in 2 packages.
#> Either pick the one you want with `::`:
#> • dplyr::mutate
#> • plyr::mutate
#> Or declare a preference with `conflicts_prefer()`:
#> • `conflicts_prefer(dplyr::mutate)`
#> • `conflicts_prefer(plyr::mutate)`
#> Backtrace:
#>     ▆
#>  1. └─conflicted (local) `<fn>`()
#>  2.   └─cli::cli_abort(...) at conflicted/R/disambiguate.R:51:4
#>  3.     └─rlang::abort(...)
```

<sup>Created on 2023-09-26 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -38,7 +38,31 @@ unique_obj <- function(name, pkgs) {
...
+        } else {
+          ""
+        }
+      },
+      error = function(e) ""
+    )
+  })
+
+  canonical_names <- unique(env_names[env_names != ""])
+  canonical_pos <- set_names(match(canonical_names, env_names), canonical_names)
```
Comment (hadley) [inline R/utils.R:60]: I'd suggest pulling this out into a `canonical_env` or similar.

File: R/utils.R [hunk 2]
```diff
@@ -38,7 +38,31 @@ unique_obj <- function(name, pkgs) {
   objs <- lapply(pkgs, getExportedValue, name)
   names(objs) <- pkgs
 
-  pkgs[!duplicated(objs)]
+  # Finding the namespace where a function is really defined
+  env_names <- map_chr(objs, function(obj) {
+    tryCatch(
```
Comment (hadley) [inline R/utils.R:43]: Can you make the `tryCatch()` more specific?

File: tests/testthat/_snaps/find.md [hunk 3]
```diff
@@ -27,3 +27,11 @@
     Message
       0 conflicts
 
+# using canonical reference
+
+    Code
+      conflict_scout(c("testthat", "dplyr", "pipe"))
+    Message
+      1 conflict
+      * `%>%()`: magrittr and pipe
```
Comment (hadley) [inline tests/testthat/_snaps/find.md:36]: This seems a bit confusing to me since you're reporting the name of a package that the user hasn't attached.

Comment (krlmlr) [reply tests/testthat/_snaps/find.md:36]: Yeah, and conflicted doesn't seem to behave well in this case too:

``` r
library(conflicted)
library(dplyr)
library(testthat)
library(fastpipe) # moodymudskipper/fastpipe

conflict_scout()
#> 4 conflicts
#> • `%>%()`: fastpipe and magrittr
#> • `filter()`: dplyr and stats
#> • `lag()`: dplyr and stats
#> • `matches()`: testthat and tidyselect
conflicts_prefer(magrittr::`%>%`)
#> [conflicted] Will prefer magrittr::`%>%` over any other package.
environment(`%>%`)
#> <environment: namespace:fastpipe>
```

<sup>Created on 2023-09-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by krlmlr** [comment]
Thanks for the guidance, it was very helpful!

**Review by hadley — APPROVED** [review]

**Pull request merged**